internal/config/config.go

```diff
@@ -0,0 +1,168 @@
+	config := Config{
+		TargetPath: targetPath,
+	}
+
+	var err error
+	config.Parameters, err = parseParameters(parametersStr)
+	if err != nil {
+		return Config{}, err
+	}
+
+	err = json.Unmarshal([]byte(permissionStr), &config.FilePermission)
+	if err != nil {
+		return Config{}, err
+	}
+
+	err = config.Validate()
+	if err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+func parseParameters(parametersStr string) (Parameters, error) {
+	var params map[string]string
+	err := json.Unmarshal([]byte(parametersStr), &params)
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	var parameters Parameters
+	parameters.VaultRoleName = params["roleName"]
+	parameters.VaultAddress = params["vaultAddress"]
+	parameters.TLSConfig.VaultCAPEM = params["vaultCAPem"]
+	parameters.TLSConfig.VaultCACertPath = params["vaultCACertPath"]
+	parameters.TLSConfig.VaultCADirectory = params["vaultCADirectory"]
+	parameters.TLSConfig.VaultTLSServerName = params["vaultTLSServerName"]
+	parameters.VaultKubernetesMountPath = params["vaultKubernetesMountPath"]
+	parameters.KubernetesServiceAccountPath = params["KubernetesServiceAccountPath"]
+	if skipTLS, ok := params["vaultSkipTLSVerify"]; ok {
+		value, err := strconv.ParseBool(skipTLS)
+		if err == nil {
+			parameters.TLSConfig.VaultSkipTLSVerify = value
+		} else {
+			return Parameters{}, err
+		}
+	}
+
+	secretsYaml := params["objects"]
+	// TODO: There is an unnecessary map under objects, instead of just directly storing an array there.
+	// Deserialisation can be simplified a fair bit if we remove it.
+	secretsMap := map[string][]string{}
+	err = yaml.Unmarshal([]byte(secretsYaml), &secretsMap)
+	if err != nil {
+		return Parameters{}, err
+	}
+	secrets, ok := secretsMap["array"]
+	if !ok {
+		return Parameters{}, errors.New("no secrets to read configured")
+	}
+	for _, s := range secrets {
+		var secret Secret
+		err = yaml.Unmarshal([]byte(s), &secret)
+		if err != nil {
+			return Parameters{}, err
+		}
+		parameters.Secrets = append(parameters.Secrets, secret)
+	}
+
+	// Set default values.
+	if parameters.VaultAddress == "" {
+		parameters.VaultAddress = defaultVaultAddress
+	}
+
+	if parameters.VaultKubernetesMountPath == "" {
+		parameters.VaultKubernetesMountPath = defaultVaultKubernetesMountPath
+	}
+	if parameters.KubernetesServiceAccountPath == "" {
+		parameters.KubernetesServiceAccountPath = defaultKubernetesServiceAccountPath
+	}
+
+	return parameters, nil
+}
+
+func (c *Config) Validate() error {
+	// Some basic validation checks.
+	if c.TargetPath == "" {
+		return errors.New("missing target path field")
+	}
+	if c.Parameters.VaultRoleName == "" {
+		return errors.Errorf("missing 'roleName' in SecretProviderClass definition")
+	}
+	certificatesConfigured := c.Parameters.TLSConfig.CertificatesConfigured()
+	if c.Parameters.TLSConfig.VaultSkipTLSVerify && certificatesConfigured {
+		return errors.New("both vaultSkipTLSVerify and TLS configuration are set")
+	}
```

Comment:
I think there are useful dev use cases where you might not want to reject this configuration. When attempting to debug TLS CA verification issues, it can be helpful to skip verify, especially in dev environments.

Reply:
I think this is still the correct thing to do in production, and in dev use cases we can edit the code to work around issues, so personally I'd prefer to leave this as-is